src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngineBackend.Linux.cs

```diff
@@ -72,9 +72,23 @@ public unsafe void EventLoop()
 
                 for (int i = 0; i < numEvents; i++)
                 {
+                    uint evts = events[i].events;
+
+                    // epoll does not play well with disconnected connection-oriented sockets, frequently
+                    // reporting spurious EPOLLHUP events. Fortunately, EPOLLHUP may be handled as an
+                    // EPOLLIN | EPOLLOUT event: the usual processing for these events will recognize and
+                    // handle the HUP condition.
+                    if ((evts & Interop.libc.EPOLLHUP) != 0)
+                    {
+                        evts = (evts & ~Interop.libc.EPOLLHUP) | Interop.libc.EPOLLIN | Interop.libc.EPOLLOUT;
+                    }
+
                     var handle = (GCHandle)events[i].data;
                     var context = (SocketAsyncContext)handle.Target;
-                    context.HandleEvents(GetSocketAsyncEvents(events[i].events));
+                    if (context != null)
```

Comment:
Consider the following sequence of events:
- `epoll_wait` or `kevent64` returns on thread 1 with an event for `SocketAsyncContext` A
- User code calls `Socket.Dispose` on thread 2, which ends up calling `SocketAsyncContext.Close()` on A, unregistering the underyling file descriptor and freeing the GC handle
- The event processing loop on thread 1 attempts to process the event for `SocketAsyncContext` A after its `GCHandle` has been freed, so `Target` is `null`.